File: src/utils/geo/createGeoJson.py
import json
from geopy.geocoders import Nominatim
from time import sleep

# Load JSON files
with open("geoData/location-based-profiles.json", "r", encoding="utf-8") as file:
    locations_data = json.load(file)

with open("researchers.json", "r", encoding="utf-8") as file:
    researcher_urls = json.load(file)

# Geocode locations
geolocator = Nominatim(user_agent="geo_locator")
location_coords = {}

for location in locations_data.keys():
    try:
        geo = geolocator.geocode(location)
        if geo:
            location_coords[location] = (geo.longitude, geo.latitude)  # GeoJSON uses [lon, lat]
        else:
            location_coords[location] = None
    except Exception as e:
        print(f"Error for {location}: {e}")
        location_coords[location] = None

    sleep(1)  # Avoid rate limits

# Build GeoJSON structure
geojson = {
    "type": "FeatureCollection",
    "features": []
}

for location, researchers in locations_data.items():
    coords = location_coords.get(location)

    if coords:  # Only include valid locations
        feature = {
            "type": "Feature",
            "properties": {
                "location": location,
                "researchers": []
            },
            "geometry": {
                "type": "Point",
                "coordinates": coords
            }
        }

        for researcher, details in researchers.items():
            works = details.get("works", [])
            url = researcher_urls.get(researcher, "No URL Found")

            feature["properties"]["researchers"].append({
                "name": researcher,
                "works": works,
                "url": url
            })

        geojson["features"].append(feature)

# Save to a GeoJSON file
with open("research_profiles.geojson", "w", encoding="utf-8") as file:
    json.dump(geojson, file, indent=2)

print("GeoJSON file created: research_profiles.geojson")

File: src/utils/geo/createGeoJson.js
const fs = require("fs");
const path = require("path");
const csv = require("csv-parser");

// Define file paths
const profilesPath = path.join(__dirname, "geoData", "location_based_profiles.json");
const urlsPath = path.join(__dirname, "geoData", "expert_url_subset.csv");
const coordsPath = path.join(__dirname, "geoData", "location_coordinates.json");
const outputPath = path.join(__dirname, "geoData", "research_profiles.geojson");

// Step 1: Load location coordinates
const locationCoordinates = JSON.parse(fs.readFileSync(coordsPath, "utf-8"));

// Step 2: Read expert URLs and store them by last name
const researcherUrls = {};

fs.createReadStream(urlsPath)
  .pipe(csv())
  .on("data", (row) => {
    const url = row["expid"]?.trim();
    const fullName = row["name"]?.trim();

    if (url && fullName) {
      const lastName = fullName.split(",")[0].toLowerCase(); // Extract last name
      researcherUrls[lastName] = url; // Store by last name
    }
  })
  .on("end", () => {
    console.log("✅ CSV Parsed.");
    generateGeoJSON();
  });

// Step 3: Read location-based profiles and generate GeoJSON
function generateGeoJSON() {
  const locationProfiles = JSON.parse(fs.readFileSync(profilesPath, "utf-8"));
  const geoJson = {
    type: "FeatureCollection",
    features: [],
  };

  for (const [location, researchers] of Object.entries(locationProfiles)) {
    const coordinates = locationCoordinates[location]; // Normalize case
    if (!coordinates) {
      console.warn(`⚠️ No coordinates found for: ${location}`);
      continue; // Skip locations without coordinates
    }

    for (const [researcherName, data] of Object.entries(researchers)) {
        const lastName = researcherName.split(",")[0].toLowerCase();
        const url = researcherUrls[lastName] || null;

      if (!url) {
        console.warn(`❌ No URL found for: ${researcherName}`);
      }

      geoJson.features.push({
        type: "Feature",
        geometry: {
          type: "Point",
          coordinates: coordinates,
        },
        properties: {
          researcher: researcherName,
          works: data.works,
          url: url,
        },
      });
    }
  }

  // Step 4: Write GeoJSON file
  fs.writeFileSync(outputPath, JSON.stringify(geoJson, null, 2));
  console.log(`🎉 GeoJSON file created: ${outputPath}`);
}
